# Notebook 1: ETL Pipeline

## ETL Pipeline for Credit Card Churn Analysis

- **Extract:** Load the credit card customer data from the dataset.
- **Transform:** Clean the data by removing duplicates, checking for and handling missing values, standardizing categorical variables, encoding categories into numeric form, and creating useful features like credit utilization ratio.
- **Load:** Save the cleaned and processed dataset for further analysis and predictive modeling to identify customers likely to churn.

This pipeline ensures your data is reliable and ready to build effective churn prediction models.

### Column Descriptions

| Column                       | Description                                                  |
| ---------------------------- | ------------------------------------------------------------ |
| **CLIENTNUM**                | Unique customer ID number                                    |
| **Attrition_Flag**           | Whether customer left the service or stayed (TARGET VARIABLE)|
| **Customer_Age**             | Age of the customer in years                                 |
| **Gender**                   | Male or Female                                               |
| **Dependent_count**          | Number of dependents (family members)                        |
| **Education_Level**          | Customer's education level                                   |
| **Marital_Status**           | Single, Married, or Divorced                                 |
| **Income_Category**          | Annual income range category                                 |
| **Card_Category**            | Type of credit card (Blue, Silver, Gold, Platinum)           |
| **Months_on_book**           | How long customer has been with the company                  |
| **Total_Relationship_Count** | Number of products customer has with the company             |
| **Months_Inactive_12_mon**   | Number of months customer was inactive in last 12 months     |
| **Contacts_Count_12_mon**    | Number of times customer contacted company in last 12 months |
| **Credit_Limit**             | Maximum credit limit allowed                                 |
| **Total_Revolving_Bal**      | Outstanding balance on the card                              |
| **Avg_Open_To_Buy**          | Average available credit remaining                           |
| **Total_Amt_Chng_Q4_Q1**     | Change in transaction amount from Q4 to Q1                   |
| **Total_Trans_Amt**          | Total transaction amount in last 12 months                   |
| **Total_Trans_Ct**           | Total transaction count in last 12 months                    |
| **Total_Ct_Chng_Q4_Q1**      | Change in transaction count from Q4 to Q1                    |
| **Avg_Utilization_Ratio**    | How much of credit limit is being used                       |
| **NB_Stay_Probabilty**       | Probability customer will stay                               |
| **NB_Churn_Probality**       | Probability customer will leave                              |

# 📥 Step 1: Extract

In this step, we will load (extract) the raw dataset into our working environment for further processing.

---

### ✅ 1️⃣ Import Required Libraries

We start by importing the necessary libraries for data handling.


In [1]:
import pandas as pd
import numpy as np

In [2]:
# Extract: Load the data
df = pd.read_csv("../data/raw_data/BankChurners.csv")


In [3]:
df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

### Key Column Renamings for Better Understanding

The columns

- `CLIENTNUM`
- `Months_on_book`
- `Total_Relationship_Count`
- `Avg_open_To_Buy`
-`Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1`  
-`Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2`  

are renamed as:  

- **Customer_ID**  
- **Tenure_Months** 
- **Products_Count** 
- **Available_Credit**
- **NB_Stay_Probability**  
- **NB_Churn_Probability**  

In [4]:
# Rename specified columns
df = df.rename(columns={
    'CLIENTNUM': 'Customer_ID',
    'Months_on_book': 'Tenure_Months',
    'Total_Relationship_Count': 'Products_Count',
    'Avg_Open_To_Buy': 'Available_Credit',
    'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1': 'NB_Stay_Probability',
    'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2': 'NB_Churn_Probability'
})


In [5]:
# Check all column names after renaming
print(df.columns)


Index(['Customer_ID', 'Attrition_Flag', 'Customer_Age', 'Gender',
       'Dependent_count', 'Education_Level', 'Marital_Status',
       'Income_Category', 'Card_Category', 'Tenure_Months', 'Products_Count',
       'Months_Inactive_12_mon', 'Contacts_Count_12_mon', 'Credit_Limit',
       'Total_Revolving_Bal', 'Available_Credit', 'Total_Amt_Chng_Q4_Q1',
       'Total_Trans_Amt', 'Total_Trans_Ct', 'Total_Ct_Chng_Q4_Q1',
       'Avg_Utilization_Ratio', 'NB_Stay_Probability', 'NB_Churn_Probability'],
      dtype='object')


In [8]:
# Display the basic information about the DataFrame:
# - Shows the total number of entries (rows)
# - Lists all column names, data types (int, float, object, etc.)
# - Shows how many non-null (non-missing) values each column has
# - Helps to quickly check for missing data and verify data types before further processing
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Customer_ID             10127 non-null  int64  
 1   Attrition_Flag          10127 non-null  object 
 2   Customer_Age            10127 non-null  int64  
 3   Gender                  10127 non-null  object 
 4   Dependent_count         10127 non-null  int64  
 5   Education_Level         10127 non-null  object 
 6   Marital_Status          10127 non-null  object 
 7   Income_Category         10127 non-null  object 
 8   Card_Category           10127 non-null  object 
 9   Tenure_Months           10127 non-null  int64  
 10  Products_Count          10127 non-null  int64  
 11  Months_Inactive_12_mon  10127 non-null  int64  
 12  Contacts_Count_12_mon   10127 non-null  int64  
 13  Credit_Limit            10127 non-null  float64
 14  Total_Revolving_Bal     10127 non-null

In [6]:
# check for missing values in each column to make sure:
# df.isna() creates a boolean DataFrame where True indicates a missing (NaN) value.
# .sum() counts the number of missing values in each column.
# This helps identify which columns have missing data and how many, 
# so you can decide how to handle them during data cleaning.
df.isna().sum()

Customer_ID               0
Attrition_Flag            0
Customer_Age              0
Gender                    0
Dependent_count           0
Education_Level           0
Marital_Status            0
Income_Category           0
Card_Category             0
Tenure_Months             0
Products_Count            0
Months_Inactive_12_mon    0
Contacts_Count_12_mon     0
Credit_Limit              0
Total_Revolving_Bal       0
Available_Credit          0
Total_Amt_Chng_Q4_Q1      0
Total_Trans_Amt           0
Total_Trans_Ct            0
Total_Ct_Chng_Q4_Q1       0
Avg_Utilization_Ratio     0
NB_Stay_Probability       0
NB_Churn_Probability      0
dtype: int64

In [7]:
# Check for duplicate rows in the DataFrame:
# df.duplicated() returns a boolean Series marking duplicates as True.
# .sum() counts how many duplicate rows exist.
# This helps identify if there are repeated records that should be removed for data quality.
df.duplicated().sum()


np.int64(0)

# The output of df.duplicated().sum() is 0,
# which means there are no duplicate rows in the dataset.
# So, no duplicate removal is needed.
# The output of df.isna().sum() shows zero missing values for all columns,
# meaning the dataset has no missing data and no imputation is needed.

## 📊 Data Summary from `df.info()`,`df.isna()`and `df.duplicated().sum()`

- **Total rows (entries):** 10,127  
- **Total columns:** 23  
- All columns have **non-null values** — no missing data found.

---

### 🧮 Data Types:

#### 🟢 Numeric Columns:

- **Integer (int64):**
  - Customer_ID
  - Customer_Age
  - Dependent_count
  - Tenure_Months
  - Products_Count
  - Months_Inactive_12_mon
  - Contacts_Count_12_mon
  - Total_Revolving_Bal
  - Total_Trans_Amt
  - Total_Trans_Ct

- **Float (float64):**
  - Credit_Limit
  - Available_Credit
  - Total_Amt_Chng_Q4_Q1
  - Total_Ct_Chng_Q4_Q1
  - Avg_Utilization_Ratio
  - NB_Stay_Probability
  - NB_Churn_Probability

#### 🟠 Categorical Columns (object):
- Attrition_Flag
- Gender
- Education_Level
- Marital_Status
- Income_Category
- Card_Category

---

### 🔍 Key Findings:

- ✅ No missing values — dataset is complete.
- ✅ No duplicates
- ✅ Data types are correct.
- 🔄 Categorical columns need encoding for machine learning models.
- 📏 Numeric columns may need scaling depending on the algorithms.
- 🚩 Some numeric columns may contain outliers that should be checked.

---

### 🚀 Conclusion:

> The dataset is clean, complete, and ready for transformation, feature engineering, encoding, and modelling.


## 📦 Step 2: Transform

In this step, we transform the data to prepare it for machine learning, basic data visualization, and storytelling in Tableau. This involves cleaning, encoding, and formatting the data to ensure it is ready for analysis and modeling.

---

In [9]:
df.head()

Customer_ID     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0    768805383  Existing Customer            45      M                3   
1    818770008  Existing Customer            49      F                5   
2    713982108  Existing Customer            51      M                3   
3    769911858  Existing Customer            40      F                4   
4    709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  Tenure_Months  \
0     High School        Married     $60K - $80K          Blue             39   
1        Graduate         Single  Less than $40K          Blue             44   
2        Graduate        Married    $80K - $120K          Blue             36   
3     High School        Unknown  Less than $40K          Blue             34   
4      Uneducated        Married     $60K - $80K          Blue             21   

   ...  Credit_Limit  Total_Revolving_Bal  Available_Credit  \
0  ...       12691.0                  777           11914.0   
1  ...        8256.0                  864            7392.0   
2  ...        3418.0                    0            3418.0   
3  ...        3313.0                 2517             796.0   
4  ...        4716.0                    0            4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  NB_Stay_Probability  NB_Churn_Probability  
0                  0.061             0.000093               0.99991  
1                  0.105             0.000057               0.99994  
2                  0.000             0.000021               0.99998  
3                  0.760             0.000134               0.99987  
4                  0.000             0.000022               0.99998  

[5 rows x 23 columns]

In [10]:
# Generate summary statistics for numerical columns in the DataFrame:
# This includes count, mean, standard deviation, min, max, and quartiles (25%, 50%, 75%).
# It helps to understand the distribution and range of the data, detect outliers,
# and get an overview of each numeric feature before further analysis.
df.describe() 


Customer_ID  Customer_Age  Dependent_count  Tenure_Months  \
count  1.012700e+04  10127.000000     10127.000000   10127.000000   
mean   7.391776e+08     46.325960         2.346203      35.928409   
std    3.690378e+07      8.016814         1.298908       7.986416   
min    7.080821e+08     26.000000         0.000000      13.000000   
25%    7.130368e+08     41.000000         1.000000      31.000000   
50%    7.179264e+08     46.000000         2.000000      36.000000   
75%    7.731435e+08     52.000000         3.000000      40.000000   
max    8.283431e+08     73.000000         5.000000      56.000000   

       Products_Count  Months_Inactive_12_mon  Contacts_Count_12_mon  \
count    10127.000000            10127.000000           10127.000000   
mean         3.812580                2.341167               2.455317   
std          1.554408                1.010622               1.106225   
min          1.000000                0.000000               0.000000   
25%          3.000000                2.000000               2.000000   
50%          4.000000                2.000000               2.000000   
75%          5.000000                3.000000               3.000000   
max          6.000000                6.000000               6.000000   

       Credit_Limit  Total_Revolving_Bal  Available_Credit  \
count  10127.000000         10127.000000      10127.000000   
mean    8631.953698          1162.814061       7469.139637   
std     9088.776650           814.987335       9090.685324   
min     1438.300000             0.000000          3.000000   
25%     2555.000000           359.000000       1324.500000   
50%     4549.000000          1276.000000       3474.000000   
75%    11067.500000          1784.000000       9859.000000   
max    34516.000000          2517.000000      34516.000000   

       Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
count          10127.000000     10127.000000    10127.000000   
mean               0.759941      4404.086304       64.858695   
std                0.219207      3397.129254       23.472570   
min                0.000000       510.000000       10.000000   
25%                0.631000      2155.500000       45.000000   
50%                0.736000      3899.000000       67.000000   
75%                0.859000      4741.000000       81.000000   
max                3.397000     18484.000000      139.000000   

       Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  NB_Stay_Probability  \
count         10127.000000           10127.000000         10127.000000   
mean              0.712222               0.274894             0.159997   
std               0.238086               0.275691             0.365301   
min               0.000000               0.000000             0.000008   
25%               0.582000               0.023000             0.000099   
50%               0.702000               0.176000             0.000181   
75%               0.818000               0.503000             0.000337   
max               3.714000               0.999000             0.999580   

       NB_Churn_Probability  
count          10127.000000  
mean               0.840003  
std                0.365301  
min                0.000420  
25%                0.999660  
50%                0.999820  
75%                0.999900  
max                0.999990

In [11]:
df.dtypes

Customer_ID                 int64
Attrition_Flag             object
Customer_Age                int64
Gender                     object
Dependent_count             int64
Education_Level            object
Marital_Status             object
Income_Category            object
Card_Category              object
Tenure_Months               int64
Products_Count              int64
Months_Inactive_12_mon      int64
Contacts_Count_12_mon       int64
Credit_Limit              float64
Total_Revolving_Bal         int64
Available_Credit          float64
Total_Amt_Chng_Q4_Q1      float64
Total_Trans_Amt             int64
Total_Trans_Ct              int64
Total_Ct_Chng_Q4_Q1       float64
Avg_Utilization_Ratio     float64
NB_Stay_Probability       float64
NB_Churn_Probability      float64
dtype: object

# Generate summary statistics for numerical columns in the DataFrame:
# This includes count, mean, standard deviation, min, max, and quartiles (25%, 50%, 75%).
# It helps to understand the distribution and range of the data, detect outliers,
# and get an overview of each numeric feature before further analysis.

In [12]:
#pd.set_option('display.max_columns', None)
df.describe()

Customer_ID  Customer_Age  Dependent_count  Tenure_Months  \
count  1.012700e+04  10127.000000     10127.000000   10127.000000   
mean   7.391776e+08     46.325960         2.346203      35.928409   
std    3.690378e+07      8.016814         1.298908       7.986416   
min    7.080821e+08     26.000000         0.000000      13.000000   
25%    7.130368e+08     41.000000         1.000000      31.000000   
50%    7.179264e+08     46.000000         2.000000      36.000000   
75%    7.731435e+08     52.000000         3.000000      40.000000   
max    8.283431e+08     73.000000         5.000000      56.000000   

       Products_Count  Months_Inactive_12_mon  Contacts_Count_12_mon  \
count    10127.000000            10127.000000           10127.000000   
mean         3.812580                2.341167               2.455317   
std          1.554408                1.010622               1.106225   
min          1.000000                0.000000               0.000000   
25%          3.000000                2.000000               2.000000   
50%          4.000000                2.000000               2.000000   
75%          5.000000                3.000000               3.000000   
max          6.000000                6.000000               6.000000   

       Credit_Limit  Total_Revolving_Bal  Available_Credit  \
count  10127.000000         10127.000000      10127.000000   
mean    8631.953698          1162.814061       7469.139637   
std     9088.776650           814.987335       9090.685324   
min     1438.300000             0.000000          3.000000   
25%     2555.000000           359.000000       1324.500000   
50%     4549.000000          1276.000000       3474.000000   
75%    11067.500000          1784.000000       9859.000000   
max    34516.000000          2517.000000      34516.000000   

       Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  \
count          10127.000000     10127.000000    10127.000000   
mean               0.759941      4404.086304       64.858695   
std                0.219207      3397.129254       23.472570   
min                0.000000       510.000000       10.000000   
25%                0.631000      2155.500000       45.000000   
50%                0.736000      3899.000000       67.000000   
75%                0.859000      4741.000000       81.000000   
max                3.397000     18484.000000      139.000000   

       Total_Ct_Chng_Q4_Q1  Avg_Utilization_Ratio  NB_Stay_Probability  \
count         10127.000000           10127.000000         10127.000000   
mean              0.712222               0.274894             0.159997   
std               0.238086               0.275691             0.365301   
min               0.000000               0.000000             0.000008   
25%               0.582000               0.023000             0.000099   
50%               0.702000               0.176000             0.000181   
75%               0.818000               0.503000             0.000337   
max               3.714000               0.999000             0.999580   

       NB_Churn_Probability  
count          10127.000000  
mean               0.840003  
std                0.365301  
min                0.000420  
25%                0.999660  
50%                0.999820  
75%                0.999900  
max                0.999990

### Known Issue: Inconsistent Visibility of Renamed Columns in `df.describe()`

**Observation:**  
After renaming columns, sometimes the renamed columns appear correctly in `df.describe()`, while at other times they are not fully displayed, even though the column names are correctly updated and visible in `df.dtypes`.

**Root Cause:**  
- This behavior is mainly related to pandas' default display settings.
- If the dataframe has many columns, pandas may truncate or limit the number of columns shown in `df.describe()` for display purposes.
- The issue is not caused by renaming itself, but rather by how much pandas chooses to display depending on session memory, notebook state, or terminal width.

**Temporary Behavior:**  
- After running multiple cells or restarting the kernel, pandas sometimes resets display settings.
- This may cause the `describe()` output to change, even though the underlying data remains correct.

**Solution Applied:**  
- Set pandas display option to always show all columns:
```python
pd.set_option('display.max_columns', None)


### Column Inconsistencies Check and Cleaning (Before Outlier Detection)

Before performing outlier detection, it is essential to ensure that all columns are clean and consistent to avoid inaccurate results. The following steps are performed:

- **Data Type Verification:** Check and correct column data types. Ensure all numerical columns are properly stored as numeric types, and categorical columns as object/string types.
- **Numeric Conversion:** Convert any numeric columns that may be incorrectly stored as object types due to data entry issues or inconsistencies.
- **Null Value Handling After Conversion:** After conversion, any non-numeric values that couldn't be converted will result in `NaN`. These are handled by filling with the median value to retain data integrity.
- **Categorical Value Inspection:** Review unique values in categorical columns to detect and correct inconsistencies such as typos, inconsistent casing, or extra spaces.
- **Standardization:** Apply standard formatting (e.g. capitalize, strip spaces) to ensure consistency across categorical values.

By cleaning these inconsistencies first, we ensure that the subsequent outlier detection and winsorization steps are accurate, reliable, and based on properly formatted data.


In [13]:
#  Check data types
print("Data Types Before Conversion:")
print(df.dtypes)

Data Types Before Conversion:
Customer_ID                 int64
Attrition_Flag             object
Customer_Age                int64
Gender                     object
Dependent_count             int64
Education_Level            object
Marital_Status             object
Income_Category            object
Card_Category              object
Tenure_Months               int64
Products_Count              int64
Months_Inactive_12_mon      int64
Contacts_Count_12_mon       int64
Credit_Limit              float64
Total_Revolving_Bal         int64
Available_Credit          float64
Total_Amt_Chng_Q4_Q1      float64
Total_Trans_Amt             int64
Total_Trans_Ct              int64
Total_Ct_Chng_Q4_Q1       float64
Avg_Utilization_Ratio     float64
NB_Stay_Probability       float64
NB_Churn_Probability      float64
dtype: object


### Data Type Review and Recommendations

**Current Situation:**

- Numeric columns are mostly of type `int64` or `float64`, which is appropriate for analysis.
- Categorical columns are currently of type `object`.
- `Customer_ID` is an integer (`int64`), but it represents a unique identifier.

**Recommendations:**

1. **Convert `Customer_ID` to string type:**

   - Treating `Customer_ID` as a string prevents accidental numeric operations and clarifies it is an identifier.

2. **Convert categorical columns to pandas `category` type:**

   - This optimizes memory usage and improves performance for categorical operations.
   - Columns to convert:  
     - `Attrition_Flag`  
     - `Gender`  
     - `Education_Level`  
     - `Marital_Status`  
     - `Income_Category`  
     - `Card_Category`  

3. **Keep numeric columns as is:**

   - The integer and float columns are correctly typed for analysis and do not require changes.

In [14]:
# Convert Customer_ID to string
df['Customer_ID'] = df['Customer_ID'].astype(str)

In [16]:
# Convert categorical columns to category dtype
categorical_cols = ['Attrition_Flag', 'Gender', 'Education_Level', 'Marital_Status', 'Income_Category', 'Card_Category']

for col in categorical_cols:
    df[col] = df[col].astype('category')


In [17]:
# After converting columns to 'category' dtype, we check the data types again
# This helps confirm that the data type changes have been successfully applied.

print(df.dtypes)


Customer_ID                 object
Attrition_Flag            category
Customer_Age                 int64
Gender                    category
Dependent_count              int64
Education_Level           category
Marital_Status            category
Income_Category           category
Card_Category             category
Tenure_Months                int64
Products_Count               int64
Months_Inactive_12_mon       int64
Contacts_Count_12_mon        int64
Credit_Limit               float64
Total_Revolving_Bal          int64
Available_Credit           float64
Total_Amt_Chng_Q4_Q1       float64
Total_Trans_Amt              int64
Total_Trans_Ct               int64
Total_Ct_Chng_Q4_Q1        float64
Avg_Utilization_Ratio      float64
NB_Stay_Probability        float64
NB_Churn_Probability       float64
dtype: object


### Data Type Validation After Conversion

After applying the necessary data type conversions (e.g. converting categorical columns to `category` dtype), we re-checked the data types to ensure that all changes were successfully applied.

**Result:**  
- All columns have the correct and expected data types.
- Numerical columns remain as `int64` or `float64`.
- Categorical columns are successfully converted to `category` dtype.

✅ The dataset is now properly formatted and ready for further processing, outlier detection, visualization, and machine learning tasks.



### Null Value Handling After Conversion for numeric columns

After converting columns to the correct data types, we perform an additional check for any `NaN` values that may have been introduced during the conversion process.

**Steps Performed:**
- Identified all numeric columns (`int64` and `float64`).
- Checked for any missing values (`NaN`) in these numeric columns.
- Filled any missing values with the **median** of each column.

**Why Median?**
- Median is less sensitive to outliers compared to mean.
- Helps maintain the original distribution of the data.
- Prevents loss of data while keeping the dataset balanced for analysis.


In [18]:
# Select numeric columns
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns

In [19]:
# Check for null values in numeric columns
print(df[numeric_cols].isna().sum())

Customer_Age              0
Dependent_count           0
Tenure_Months             0
Products_Count            0
Months_Inactive_12_mon    0
Contacts_Count_12_mon     0
Credit_Limit              0
Total_Revolving_Bal       0
Available_Credit          0
Total_Amt_Chng_Q4_Q1      0
Total_Trans_Amt           0
Total_Trans_Ct            0
Total_Ct_Chng_Q4_Q1       0
Avg_Utilization_Ratio     0
NB_Stay_Probability       0
NB_Churn_Probability      0
dtype: int64


### Null Value Check in Numeric Columns

Before applying imputation, we performed a null value check on all numeric columns using `df[numeric_cols].isna().sum()`:

- This ensures that any missing values generated during type conversion or data ingestion are identified.
- If missing values are found, appropriate imputation (e.g., median) will be applied to retain data quality.
- In this case,we have no numeric columns with nulls.


### Handling Null Values in Categorical Columns

Before filling missing values, we first check if there are any null values in the categorical columns (`object` and `category` types).

**Why This Step Is Important:**

- Categorical columns may contain missing values that can affect:
  - Encoding
  - Machine learning algorithms
  - Visualizations

- Most machine learning algorithms cannot handle missing values in categorical features.
- Cleaning these columns ensures better model performance and avoids unexpected errors.

**Strategy:**

1. Check for null values in all categorical columns.
2. If null values are found, fill them using either:
   - The mode (most frequent value), or
   - A placeholder value (e.g., `'Unknown'`), depending on business logic.

**In this project, we will use the mode to retain the existing data distribution.**

In [20]:
# Select categorical columns (object and category types)
categorical_cols = df.select_dtypes(include=['object', 'category']).columns

In [21]:
# Check for null values
print(df[categorical_cols].isna().sum())

Customer_ID        0
Attrition_Flag     0
Gender             0
Education_Level    0
Marital_Status     0
Income_Category    0
Card_Category      0
dtype: int64


### Null Value Check in Categorical Columns

After checking the categorical columns (`object` and `category` types), no null values were found.  

This confirms the data consistency in categorical features, allowing us to proceed with encoding and modeling without needing to handle missing values in these columns.


### Categorical Value Inspection

To ensure data quality and consistency, we inspect the unique values in all categorical columns. This helps to identify issues such as:

- Typos (e.g., 'Male' vs 'male')
- Inconsistent casing (e.g., 'Single' vs 'single')
- Leading/trailing spaces or extra characters

Detecting and correcting these inconsistencies early helps improve the accuracy of encoding, modeling, and reporting.


In [23]:
# Step 1: Identify categorical columns (object or category dtype)
categorical_cols = df.select_dtypes(include=['object', 'category']).columns

# Step 2: Review unique values for each categorical column to spot inconsistencies
for col in categorical_cols:
    print(f"Unique values in '{col}':")
    print(df[col].unique())
    print("-" * 40)


Unique values in 'Customer_ID':
['768805383' '818770008' '713982108' ... '716506083' '717406983'
 '714337233']
----------------------------------------
Unique values in 'Attrition_Flag':
['Existing Customer', 'Attrited Customer']
Categories (2, object): ['Attrited Customer', 'Existing Customer']
----------------------------------------
Unique values in 'Gender':
['M', 'F']
Categories (2, object): ['F', 'M']
----------------------------------------
Unique values in 'Education_Level':
['High School', 'Graduate', 'Uneducated', 'Unknown', 'College', 'Post-Graduate', 'Doctorate']
Categories (7, object): ['College', 'Doctorate', 'Graduate', 'High School', 'Post-Graduate', 'Uneducated', 'Unknown']
----------------------------------------
Unique values in 'Marital_Status':
['Married', 'Single', 'Unknown', 'Divorced']
Categories (4, object): ['Divorced', 'Married', 'Single', 'Unknown']
----------------------------------------
Unique values in 'Income_Category':
['$60K - $80K', 'Less than $40K',

### Categorical Value Inspection

We inspected the unique values in each categorical column to identify any inconsistencies such as typos, inconsistent casing, or unexpected categories. This is an important step to ensure data quality before encoding and modeling.

**Observations:**

- The categorical columns (`Gender`, `Education_Level`, `Marital_Status`, `Income_Category`, `Card_Category`, `Attrition_Flag`) have consistent and expected values.
- Categories like `Unknown` are used to indicate missing or unknown data in some columns, which is consistent across the dataset.
- No misspelled or unexpected category values were found.


### Handling "Unknown" Categories in Categorical Data

The presence of "Unknown" values in categorical columns indicates missing or unspecified information. Instead of removing these rows, which could lead to data loss, we can:

- **Treat "Unknown" as a Separate Category:**  
  This approach keeps all data intact and allows models to learn patterns associated with missing information.

- **Impute with Most Frequent Category (Mode):**  
  If domain knowledge suggests, replace "Unknown" with the most common category to reduce ambiguity.

- **Use Domain-Specific Rules:**  
  Sometimes, "Unknown" can be mapped to a meaningful category based on business rules.

**Recommendation:**  
In this project, we will **keep "Unknown" as a valid category** to preserve data completeness and avoid losing valuable information.

This approach ensures models are aware of missing data patterns without data removal.



In [27]:
# List of categorical columns that contain 'Unknown' values
cat_cols_with_unknown = ['Education_Level', 'Marital_Status', 'Income_Category']

for col in cat_cols_with_unknown:
    # Ensure the column is of type 'category'
    df[col] = df[col].astype('category')
    
# Check if 'Unknown' is already a category
    if 'Unknown' not in df[col].cat.categories:
        # If not, add 'Unknown' explicitly as a category
        df[col] = df[col].cat.add_categories('Unknown')
    
    # Print value counts including 'Unknown' to verify it's treated as a category
    print(f"Value counts for {col}:")
    print(df[col].value_counts(dropna=False))
    print('-' * 40)

Value counts for Education_Level:
Education_Level
Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: count, dtype: int64
----------------------------------------
Value counts for Marital_Status:
Marital_Status
Married     4687
Single      3943
Unknown      749
Divorced     748
Name: count, dtype: int64
----------------------------------------
Value counts for Income_Category:
Income_Category
Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: count, dtype: int64
----------------------------------------


### Handling 'Unknown' as a Separate Category

We have ensured that `'Unknown'` is treated as a distinct category in relevant categorical columns. This allows us to keep these entries without removing data, preserving information and improving model and visualization accuracy.

By explicitly including `'Unknown'` as a category, we avoid errors during encoding and enable the model to learn from these unknown or missing value groups.


### Standardization of Categorical Values

To maintain consistency across categorical columns, we apply standard formatting by:

- Removing any leading or trailing spaces from each value.
- Capitalizing the first letter of each word (title case).

This ensures uniformity in categorical data, which is critical for accurate analysis, visualization, and machine learning.

After standardization, we convert the columns back to the `category` datatype to optimize memory usage and improve performance.



In [28]:
# List of categorical columns to standardize
cat_columns = ['Education_Level', 'Marital_Status', 'Income_Category', 'Card_Category', 'Gender', 'Attrition_Flag']

for col in cat_columns:
    # Remove leading/trailing spaces and capitalize each value for consistency
    df[col] = df[col].str.strip().str.title()


In [30]:
# Convert column back to categorical type after standardization
df[col] = df[col].astype('category')

### Why Do We Convert Back to `category` Type?

During string operations, pandas automatically converts `category` columns to `object` type.  
We now convert them back to `category` to:

- ✅ Optimize memory usage (store unique categories efficiently)
- ✅ Speed up computations for grouping, analysis, and encoding
- ✅ Preserve that these columns contain fixed discrete categories
- ✅ Ensure compatibility with machine learning algorithms and pipelines



In [31]:
df.head(2)

Customer_ID     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0   768805383  Existing Customer            45      M                3   
1   818770008  Existing Customer            49      F                5   

  Education_Level Marital_Status Income_Category Card_Category  Tenure_Months  \
0     High School        Married     $60K - $80K          Blue             39   
1        Graduate         Single  Less Than $40K          Blue             44   

   ...  Credit_Limit  Total_Revolving_Bal  Available_Credit  \
0  ...       12691.0                  777           11914.0   
1  ...        8256.0                  864            7392.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   

   Avg_Utilization_Ratio  NB_Stay_Probability  NB_Churn_Probability  
0                  0.061             0.000093               0.99991  
1                  0.105             0.000057               0.99994  

[2 rows x 23 columns]

## Handling Outliers Without Removing Using Winsorization

To avoid losing data points while reducing the influence of extreme values (outliers), we apply Winsorization:

- Calculate outlier thresholds for each numeric column using the Interquartile Range (IQR) method.
- Cap values below the lower threshold to the lower bound.
- Cap values above the upper threshold to the upper bound.
- This method retains all data but limits the effect of outliers on analysis and modeling.


In [32]:
# List of numeric columns to check and Winsorize
numeric_cols = df.select_dtypes(include=['int64', 'float64']).columns

for col in numeric_cols:
    # Calculate Q1 (25th percentile) and Q3 (75th percentile)
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1
    
    # Calculate lower and upper bounds for outliers
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    # Apply Winsorization by capping values outside the bounds
    df[col] = df[col].clip(lower=lower_bound, upper=upper_bound)
    
    # Optional: print summary of changes
    print(f"{col}: capped values below {lower_bound:.2f} and above {upper_bound:.2f}")

# Now df has outliers capped instead of removed


Customer_Age: capped values below 24.50 and above 68.50
Dependent_count: capped values below -2.00 and above 6.00
Tenure_Months: capped values below 17.50 and above 53.50
Products_Count: capped values below 0.00 and above 8.00
Months_Inactive_12_mon: capped values below 0.50 and above 4.50
Contacts_Count_12_mon: capped values below 0.50 and above 4.50
Credit_Limit: capped values below -10213.75 and above 23836.25
Total_Revolving_Bal: capped values below -1778.50 and above 3921.50
Available_Credit: capped values below -11477.25 and above 22660.75
Total_Amt_Chng_Q4_Q1: capped values below 0.29 and above 1.20
Total_Trans_Amt: capped values below -1722.75 and above 8619.25
Total_Trans_Ct: capped values below -9.00 and above 135.00
Total_Ct_Chng_Q4_Q1: capped values below 0.23 and above 1.17
Avg_Utilization_Ratio: capped values below -0.70 and above 1.22
NB_Stay_Probability: capped values below -0.00 and above 0.00
NB_Churn_Probability: capped values below 1.00 and above 1.00


### Outlier Capping Summary

Using the Interquartile Range (IQR) method, we identified the lower and upper bounds for each numeric column. Values outside these bounds were capped to limit the effect of extreme outliers while retaining all data points.

| Column                  | Capped Lower Bound | Capped Upper Bound |
|-------------------------|--------------------|--------------------|
| **Customer_Age**            | 24.50              | 68.50              |
| **Dependent_count**         | -2.00              | 6.00               |
| **Tenure_Months**           | 17.50              | 53.50              |
| **Products_Count**          | 0.00               | 8.00               |
| **Months_Inactive_12_mon**  | 0.50               | 4.50               |
| **Contacts_Count_12_mon**   | 0.50               | 4.50               |
| **Credit_Limit**            | -10213.75          | 23836.25           |
| **Total_Revolving_Bal**     | -1778.50           | 3921.50            |
| **Available_Credit**        | -11477.25          | 22660.75           |
| **Total_Amt_Chng_Q4_Q1**    | 0.29               | 1.20               |
| **Total_Trans_Amt**         | -1722.75           | 8619.25            |
| **Total_Trans_Ct**          | -9.00              | 135.00             |
| **Total_Ct_Chng_Q4_Q1**     | 0.23               | 1.17               |
| **Avg_Utilization_Ratio**   | -0.70              | 1.22               |
| **NB_Stay_Probability**     | -0.00              | 0.00               |
| **NB_Churn_Probability**    | 1.00               | 1.00               |

> **Note:** Negative lower bounds for some columns indicate theoretical limits based on IQR calculations; actual data values were capped accordingly, typically at the minimum or zero where negative values are not feasible.

> Theoretical limits are the boundary values calculated from formulas to detect outliers, but sometimes these limits can be unrealistic (like negative ages), so we adjust them to fit real-world data.

---

This approach preserves all data while minimizing the influence of extreme outliers, improving model robustness and stability.

In [33]:
# After completing cleaning:
df_cleaned = df.copy()


### Encoding Categorical Variables

Machine learning models require numeric input. Thus, categorical variables must be converted to numeric formats.

- **One-Hot Encoding** is used for nominal categorical variables without order.
- **Label Encoding** is applied to ordinal variables with inherent order.

We will encode the following columns:
- `Attrition_Flag`
- `Gender`
- `Education_Level`
- `Marital_Status`
- `Income_Category`
- `Card_Category`

Encoding ensures the model can interpret categorical information effectively.

### Creating Encoded DataFrame for Machine Learning

- We preserve the original cleaned dataframe (`df_cleaned`) for descriptive analysis, Correlation Analysis ,segmentation, and basic visualization.
- We create a separate encoded dataframe (`df_encoded`) to transform categorical columns into numeric values suitable for machine learning algorithms.
- This approach allows flexible use of both datasets depending on the analysis stage.


### Encoding Categorical Variables

Machine learning models require numeric input. Thus, categorical variables must be converted to numeric formats.

- **Nominal Variables:** Categories without any intrinsic order or ranking.  
  Examples: `Gender` (Male/Female), `Card_Category` (Blue, Silver, Gold, Platinum).

- **Ordinal Variables:** Categories with a meaningful order or ranking.  
  Examples: `Education_Level` (High School < College < Graduate < Post-Graduate < Doctorate).

- **One-Hot Encoding** is used for **nominal** categorical variables without order.  
- **Label Encoding** is applied to **ordinal** variables with inherent order.

We will encode the following columns:  
- `Attrition_Flag` (nominal)  
- `Gender` (nominal)  
- `Education_Level` (ordinal)  
- `Marital_Status` (nominal)  
- `Income_Category` (ordinal)  
- `Card_Category` (nominal)  

Encoding ensures the model can interpret categorical information effectively.



In [34]:
!pip install scikit-learn


  Using cached joblib-1.5.1-py3-none-any.whl.metadata (5.6 kB)
  Using cached threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 10.3 MB/s eta 0:00:0000:010:01
Using cached joblib-1.5.1-py3-none-any.whl (307 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.7/38.7 MB 9.7 MB/s eta 0:00:00:00:0100:01
Using cached threadpoolctl-3.6.0-py3-none-any.whl (18 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4/4 [scikit-learn] [scikit-learn]


In [35]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder

# Make a copy for safety
df_encoded = df_cleaned.copy()

### Label Encoding for binary and ordinal columns

# 1. Attrition_Flag (binary: 'Existing Customer', 'Attrited Customer')
df_encoded['Attrition_Flag'] = df_encoded['Attrition_Flag'].map({'Existing Customer': 0, 'Attrited Customer': 1})

# 2. Gender (binary: 'M', 'F')
df_encoded['Gender'] = df_encoded['Gender'].map({'M': 0, 'F': 1})

# 3. Education_Level (ordinal)
edu_order = ['Uneducated', 'High School', 'College', 'Graduate', 'Post-Graduate', 'Doctorate', 'Unknown']
edu_encoder = LabelEncoder()
edu_encoder.fit(edu_order)
df_encoded['Education_Level'] = edu_encoder.transform(df_encoded['Education_Level'])

# 4. Income_Category (ordinal)
inc_order = ['Less Than $40K', '$40K - $60K', '$60K - $80K', '$80K - $120K', '$120K +', 'Unknown']
inc_encoder = LabelEncoder()
inc_encoder.fit(inc_order)
df_encoded['Income_Category'] = inc_encoder.transform(df_encoded['Income_Category'])

### One-Hot Encoding for nominal columns

# 5. Marital_Status
df_encoded = pd.get_dummies(df_encoded, columns=['Marital_Status'], prefix='Marital')

# 6. Card_Category
df_encoded = pd.get_dummies(df_encoded, columns=['Card_Category'], prefix='Card')

# Check the result
df_encoded.head()


Customer_ID Attrition_Flag  Customer_Age  Gender  Dependent_count  \
0   768805383              0          45.0       0                3   
1   818770008              0          49.0       1                5   
2   713982108              0          51.0       0                3   
3   769911858              0          40.0       1                4   
4   709106358              0          40.0       0                3   

   Education_Level  Income_Category  Tenure_Months  Products_Count  \
0                3                2           39.0               5   
1                2                4           44.0               6   
2                2                3           36.0               4   
3                3                4           34.0               3   
4                5                2           21.0               5   

   Months_Inactive_12_mon  ...  NB_Stay_Probability  NB_Churn_Probability  \
0                     1.0  ...             0.000093               0.99991   
1                     1.0  ...             0.000057               0.99994   
2                     1.0  ...             0.000021               0.99998   
3                     4.0  ...             0.000134               0.99987   
4                     1.0  ...             0.000022               0.99998   

   Marital_Divorced  Marital_Married  Marital_Single  Marital_Unknown  \
0             False             True           False            False   
1             False            False            True            False   
2             False             True           False            False   
3             False            False           False             True   
4             False             True           False            False   

   Card_Blue  Card_Gold  Card_Platinum  Card_Silver  
0       True      False          False        False  
1       True      False          False        False  
2       True      False          False        False  
3       True      False          False        False  
4       True      False          False        False  

[5 rows x 29 columns]

In [36]:
print(df_encoded['Income_Category'].unique())

[2 4 3 1 0 5]
